### PR TITLE
Cleans up code and refactors tests for single qubit unitary decompositions

### PR DESCRIPTION
**Pull request by astralcai**
**Context:**

As I am implementing the XZX single qubit decomposition (https://github.com/PennyLaneAI/pennylane/pull/4862), I noticed some issues with the the code and the tests for single qubit unitary decompositions.

**Description of the Change:**

Cleans up ```transforms/decompositions/single_qubit_unitary.py```. 

- Removes duplicate code,
- Fixes doc examples that were untrue
- Adds support for returning global phase for the `rot` decomposition.
- Normalize all rotation angles to the range $[0, 4\pi]$

Refactors ```tests/transforms/test_decompositions.py```. 

- Generalizes logic for obtaining decompositions and asserting their correctness such that all types of decompositions share the same assertion logic. Removes duplicate code.
- Adds testing for when ```return_global_phase=False``` as well as when ```return_global_phase=True```
- Update test cases to use analytical instead of numerical values when possible.

**Benefits:**

Cleaner code.

**Related GitHub Issues:**

https://github.com/PennyLaneAI/pennylane/issues/4868
https://github.com/PennyLaneAI/pennylane/pull/4862

### CODE REVIEW

**Comment by rmoyard** [comment]
Hi @astralcai  thanks a lot for this follow up PR. I am not sure it is possible to change the base of this PR because you are external, but that would be nice. Currently it is hard to distinguish your two PRs. If it is not possible, I will wait for the first one to be merged before reviewing.

**Comment by astralcai** [comment]
> Hi @astralcai thanks a lot for this follow up PR. I am not sure it is possible to change the base of this PR because you are external, but that would be nice. Currently it is hard to distinguish your two PRs. If it is not possible, I will wait for the first one to be merged before reviewing.

@rmoyard Hi, I have made another update to this PR, now normalizing all rotation angles to $[-2\pi, 2\pi]$ for consistency and easy of use (previously different types composition deals with this differently). I have also updated all test cases for single-qubit unitary decompositions to use analytical values as expected parameters when possible.

**Comment by rmoyard** [comment]
Hi @astralcai ! Once you solve the conflicts I will review this PR. One thing about angles for rotation, our canonical range is `[0, 4pi]`, same as Numpy. So it would be best to remove your angle change from this PR. Thanks :+1:

**Comment by astralcai** [comment]
> Hi @astralcai ! Once you solve the conflicts I will review this PR. One thing about angles for rotation, our canonical range is `[0, 4pi]`, same as Numpy. So it would be best to remove your angle change from this PR. Thanks 👍

@rmoyard In this case, I will update the code for the `rot` decomposition to normalize its rotation angles to the `[0, 4pi]` range instead. 

Can I ask why this range is chosen?

**Comment by astralcai** [comment]
> > Hi @astralcai ! Once you solve the conflicts I will review this PR. One thing about angles for rotation, our canonical range is `[0, 4pi]`, same as Numpy. So it would be best to remove your angle change from this PR. Thanks 👍
> 
> @rmoyard In this case, I will update the code for the `rot` decomposition to normalize its rotation angles to the `[0, 4pi]` range instead.
> 
> Can I ask why this range is chosen?

@rmoyard I have updated the PR to remove the angle change. It is ready for review now.

**Comment by rmoyard** [comment]
Thank you I will take a look between today and tomorrow. From what I remember, we wanted to follow the convention by numpy.mod and % where the rest is not negative. We thought it was a better software friendly convention compared to the other one which comes more from physics.

File: doc/releases/changelog-dev.md
Comment (rmoyard) [inline doc/releases/changelog-dev.md:91]: I would move this one to breaking change.

File: pennylane/transforms/decompositions/single_qubit_unitary.py
Comment (rmoyard) [inline pennylane/transforms/decompositions/single_qubit_unitary.py:157]: Thanks for this change it looks great!

File: tests/ops/qubit/test_matrix_ops.py
Comment (rmoyard) [inline tests/ops/qubit/test_matrix_ops.py:270]: Thanks for cleaning up the values.

File: tests/transforms/test_unitary_to_rot.py
Comment (rmoyard) [inline tests/transforms/test_unitary_to_rot.py:34]: :+1:

File: doc/releases/changelog-dev.md
Comment (rmoyard) [inline doc/releases/changelog-dev.md:91]: I would add that global phase is available for rot in the improvement though.

File: pennylane/transforms/decompositions/single_qubit_unitary.py
Comment (rmoyard) [inline pennylane/transforms/decompositions/single_qubit_unitary.py:85]: If I am not wrong you can also add the normalization of angle in the helper function.

**Review by rmoyard — COMMENTED** [review]
Thanks for cleaning up the tests and the changes to `rot`. It looks great, I have left some comments. Happy to approve after that :100:

**Review by lillian542 — APPROVED** [review]
I don't have anything to add to @rmoyard's feedback. Thanks @astralcai, this looks great! 🚀

**Review by rmoyard — APPROVED** [review]
Thanks @astralcai, feel free to merge the PR :100:

**Comment by astralcai** [comment]
> Thanks @astralcai, feel free to merge the PR 💯

Hi, thank you. I don't think I have permission to merge this PR.

**Pull request merged**
